File: app/src/main/java/me/devsaki/hentoid/viewholders/SiteItem.java
package me.devsaki.hentoid.viewholders;

import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.drag.IExtendedDraggable;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter.utils.DragDropUtil;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.enums.Site;
import me.devsaki.hentoid.util.ThemeHelper;

public class SiteItem extends AbstractItem<SiteItem.SiteViewHolder> implements IExtendedDraggable {

    private final Site site;
    private final boolean showHandle;
    private final ItemTouchHelper touchHelper;

    public SiteItem(Site site, boolean selected, ItemTouchHelper touchHelper) {
        this.site = site;
        this.setSelected(selected);
        this.showHandle = true;
        this.touchHelper = touchHelper;
    }

    public SiteItem(Site site, boolean selected, boolean showHandle) {
        this.site = site;
        this.showHandle = showHandle;
        this.setSelected(selected);
        this.touchHelper = null;
    }

    public Site getSite() {
        return site;
    }


    @NotNull
    @Override
    public SiteViewHolder getViewHolder(@NotNull View view) {
        return new SiteViewHolder(view);
    }

    @Override
    public int getLayoutRes() {
        return R.layout.item_drawer_edit;
    }

    @Override
    public int getType() {
        return R.id.drawer_edit;
    }

    @Override
    public boolean isDraggable() {
        return true;
    }

    @Nullable
    @Override
    public ItemTouchHelper getTouchHelper() {
        return touchHelper;
    }

    @Nullable
    @Override
    public View getDragView(@NotNull RecyclerView.ViewHolder viewHolder) {
        if (viewHolder instanceof SiteViewHolder)
            return ((SiteViewHolder) viewHolder).dragHandle;
        else return null;
    }


    static class SiteViewHolder extends FastAdapter.ViewHolder<SiteItem> implements IDraggableViewHolder {

        private final View rootView;
        private final ImageView dragHandle;
        private final ImageView icon;
        private final TextView title;
        private final CheckBox chk;

        SiteViewHolder(View view) {
            super(view);
            rootView = view;
            dragHandle = view.findViewById(R.id.drawer_item_handle);
            icon = view.findViewById(R.id.drawer_item_icon);
            title = view.findViewById(R.id.drawer_item_txt);
            chk = view.findViewById(R.id.drawer_item_chk);
        }

        @Override
        public void bindView(@NotNull SiteItem item, @NotNull List<?> list) {
            dragHandle.setVisibility(item.showHandle ? View.VISIBLE : View.GONE);
            if (item.showHandle) DragDropUtil.bindDragHandle(this, item);
            title.setText(item.site.getDescription());
            icon.setImageResource(item.site.getIco());
            chk.setChecked(item.isSelected());
            chk.setOnCheckedChangeListener((v, b) -> item.setSelected(b));
        }

        @Override
        public void unbindView(@NotNull SiteItem item) {
            chk.setOnCheckedChangeListener(null);
        }

        @Override
        public void onDragged() {
            rootView.setBackgroundColor(ThemeHelper.getColor(rootView.getContext(), R.color.white_opacity_25));
        }

        @Override
        public void onDropped() {
            rootView.setBackgroundColor(ThemeHelper.getColor(rootView.getContext(), R.color.transparent));
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/viewholders/SiteItem.kt
package me.devsaki.hentoid.viewholders

import android.view.View
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.drag.IExtendedDraggable
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.utils.DragDropUtil.bindDragHandle
import me.devsaki.hentoid.R
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.util.getThemedColor

class SiteItem : AbstractItem<SiteItem.ViewHolder>, IExtendedDraggable<SiteItem.ViewHolder> {

    val site: Site
    private var showHandle = false
    private var mTouchHelper: ItemTouchHelper? = null

    constructor(site: Site, selected: Boolean, touchHelper: ItemTouchHelper) : super() {
        this.site = site
        isSelected = selected
        showHandle = true
        mTouchHelper = touchHelper
    }

    constructor(site: Site, selected: Boolean, showHandle: Boolean) : super() {
        this.site = site
        this.showHandle = showHandle
        isSelected = selected
        mTouchHelper = null
    }

    override val type: Int
        get() = R.id.drawer_edit
    override val layoutRes: Int
        get() = R.layout.item_drawer_edit
    override val isDraggable: Boolean
        get() = true
    override val touchHelper: ItemTouchHelper?
        get() = mTouchHelper

    override fun getDragView(viewHolder: ViewHolder): View {
        return viewHolder.dragHandle
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    class ViewHolder internal constructor(view: View) : FastAdapter.ViewHolder<SiteItem>(view),
        IDraggableViewHolder {
        private val rootView: View = view
        val dragHandle: ImageView = view.findViewById(R.id.drawer_item_handle)
        private val icon: ImageView = view.findViewById(R.id.drawer_item_icon)
        private val title: TextView = view.findViewById(R.id.drawer_item_txt)
        private val chk: CheckBox = view.findViewById(R.id.drawer_item_chk)

        override fun bindView(item: SiteItem, payloads: List<Any>) {
            dragHandle.visibility = if (item.showHandle) View.VISIBLE else View.GONE
            @Suppress("UNCHECKED_CAST")
            if (item.showHandle) bindDragHandle(
                this,
                item as IExtendedDraggable<RecyclerView.ViewHolder>
            )
            title.text = item.site.description
            icon.setImageResource(item.site.ico)
            chk.isChecked = item.isSelected
            chk.setOnCheckedChangeListener { _, b -> item.isSelected = b }
        }

        override fun unbindView(item: SiteItem) {
            chk.setOnCheckedChangeListener(null)
        }

        override fun onDragged() {
            rootView.setBackgroundColor(rootView.context.getThemedColor(R.color.white_opacity_25))
        }

        override fun onDropped() {
            rootView.setBackgroundColor(rootView.context.getThemedColor(R.color.transparent))
        }
    }
}